### ISSUE TITLE
ReduceLROnPlateau detects a plateau during a steady decrease after a spike

### ISSUE THREAD

**Issue by SamPruden** [pytorch/pytorch#30873]
I'm not sure if this is best described as a bug report, feature request, or even documentation clarification. I'm describing an unintuitive and undesired behaviour, but not necessarily a wrong behaviour.

The name and documentation of `ReduceLROnPlateau` suggests that it will step the LR when the loss is no longer decreasing, but that's not how it behaves. It actually steps the LR when the loss is not below the best it has ever achieved. This causes problems if the loss spikes up during training then continues decreasing from that spike.

Here's a loss graph where this happens:
![image](https://user-images.githubusercontent.com/19550501/70337933-e01b9180-1843-11ea-976a-422172e2a739.png)

Everything after step ~250 is treated as a plateau by `ReduceLROnPlateau` because it's not below the lowest point, despite the fact that it's clearly trending down and not a plateau in the conventional sense of the word. In this example, this stalls the training because the LR is repeatedly decimated and it never gets a chance to get back below the lowest point. With the scheduler disabled, it keeps decreasing rapidly and quickly gets below the previous lowest point with no further spikes.

The documentation states:
> Reduce learning rate when a metric has stopped improving. Models often benefit from reducing the learning rate by a factor of 2-10 once learning stagnates. This scheduler reads a metrics quantity and if no improvement is seen for a ‘patience’ number of epochs, the learning rate is reduced.

I don't think the behaviour and the description match. The loss is clearly improving at the point when the LR is being reduced.

I'd like the plateau detection to be more sensible about not triggering when there's a downward trend. That's tricky, because a cycle of spike, gentle decrease, spike, gentle decrease should be detected as a plateau, but a one off case shouldn't. Perhaps one way to approach this would be to give `best` a maximum age, so that it's only ever looking at the most recent `n` steps. I wonder if `patience` would be a good default value of `n`?

**Comment by SamPruden** [comment]
I should add that in my particular case I've solved this by just not stepping the scheduler until step 1000, so this isn't causing me any real trouble at the moment. It could be pretty painful if this happened deep into a long training session, though.

**Comment by SamPruden** [comment]
I actually found that this repeatedly caused problems for me and I've had to move away from this scheduler as a result. The detected plateaus are often areas of noisy loss, and `ReduceLROnPlateau` kicks in and leaves me stranded at the top of a spike, then the LR collapses and training stalls.

#2478 is another attempt to solve a similar problem by adding backtracking, and that got as far as a PR in #2544 but progress there seems to have stalled in 2017.

I'm using a simple manual per step schedule for now, but that's a bit fiddly (it needs a lot of adjustment) and it would be great if the automated approaches were more stable.

**Comment by vincentqb** [comment]
Can you provide your version and steps to reproduce?

**Comment by SamPruden** [comment]
> Can you provide your version and steps to reproduce?

My version is 1.3.1.

I can try, but I'll need to provide a model that has the right behaviour when training to trigger this, so I can't easily do a tiny toy repro. Is that really needed? The behaviour I'm describing seems pretty obvious from the algorithm.

It's plain to see from [the source](https://github.com/pytorch/pytorch/blob/23483406aab765c10a74a3466ea8c17322b997d6/torch/optim/lr_scheduler.py#L492) that the condition for plateaus is simply `patience` steps without the current value being better than `best`, and that `best` is always the best value that has ever been seen. It's obvious that if the loss jumps up above the best, it will be considered a plateau with no consideration given to the trend.

I'm happy to try to repro this, but it might get messy. I might be able to cheat with a simulation model.

**Comment by SamPruden** [comment]
Here's a simulated repro. If you run this, you'll see that it never recovers from the spike at 50, despite the fact that it's learning steadily.

```python
import torch

class ToyOptimiser(torch.optim.Optimizer):
  def __init__(self, lr):
    # The scheduler just updates this
    self.param_groups = [{"lr": lr}]

optimiser = ToyOptimiser(0.01)
scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimiser)

# No actual model, but we can simulate training by just decreasing a loss.
model_loss = 1

for i in range(100):
  cur_lr = optimiser.param_groups[0]["lr"] # As set by the scheduler
  model_loss -= cur_lr # Simulate learning
  if i == 50: model_loss += 0.5 # Simulate a spike 
  scheduler.step(model_loss) # Step the real scheduler which we're testing

  print(f"{i} - {model_loss} - {cur_lr}")
```

**Comment by SamPruden** [comment]
Two years later I've just run into this annoyance again. Whether or not the behaviour is desired, it seems pretty clear that what it actually does is different from what the name and docs claim.
